internal/app/component_instance/component_instance_handler_events.go

```diff
@@ -158,3 +163,147 @@ type ListContextsEvent struct {
 func (e *ListContextsEvent) Name() event.EventName {
 	return ListContextsEventName
 }
+
+// OnComponentInstanceCreateAuthz is a handler for the CreateComponentInstanceEvent
+// It creates an OpenFGA relation tuple for the component instance and the current user
+func OnComponentInstanceCreateAuthz(db database.Database, e event.Event, authz openfga.Authorization) {
+	defaultPrio := db.GetDefaultIssuePriority()
+	defaultRepoName := db.GetDefaultRepositoryName()
+
+	l := logrus.WithFields(logrus.Fields{
+		"event":             "OnComponentInstanceCreateAuthz",
+		"payload":           e,
+		"default_priority":  defaultPrio,
+		"default_repo_name": defaultRepoName,
+	})
+
+	if createEvent, ok := e.(*CreateComponentInstanceEvent); ok {
+		instanceId := strconv.FormatInt(createEvent.ComponentInstance.Id, 10)
+		serviceId := strconv.FormatInt(createEvent.ComponentInstance.ServiceId, 10)
+		componentVersionId := strconv.FormatInt(createEvent.ComponentInstance.ComponentVersionId, 10)
+		userId := authz.GetCurrentUser()
+
+		rlist := []openfga.RelationInput{
+			{
+				UserType:   "role",
+				UserId:     openfga.UserId(userId),
+				Relation:   "role",
+				ObjectType: "component_instance",
+				ObjectId:   openfga.ObjectId(instanceId),
+			},
+			{
+				UserType:   "service",
+				UserId:     openfga.UserId(serviceId),
+				Relation:   "related_service",
+				ObjectType: "component_instance",
+				ObjectId:   openfga.ObjectId(instanceId),
+			},
+			{
+				UserType:   "component_instance",
+				UserId:     openfga.UserId(instanceId),
+				Relation:   "role",
+				ObjectType: "component_version",
+				ObjectId:   openfga.ObjectId(componentVersionId),
+			},
+		}
+
+		for _, rel := range rlist {
+			authz.AddRelation(rel)
+		}
+	} else {
+		l.Error("Wrong event")
+	}
+}
+
+// OnComponentInstanceUpdateAuthz is a handler for the UpdateComponentInstanceEvent
+// Fields that can be updated in Component Instance which affect tuple relations include:
+// componentinstance_component_version_id
+// componentinstance_service_id
+func OnComponentInstanceUpdateAuthz(db database.Database, e event.Event, authz openfga.Authorization) {
+	defaultPrio := db.GetDefaultIssuePriority()
+	defaultRepoName := db.GetDefaultRepositoryName()
+
+	l := logrus.WithFields(logrus.Fields{
+		"event":             "OnComponentInstanceUpdateAuthz",
+		"payload":           e,
+		"default_priority":  defaultPrio,
+		"default_repo_name": defaultRepoName,
+	})
+
+	if updateEvent, ok := e.(*UpdateComponentInstanceEvent); ok {
+		instanceId := strconv.FormatInt(updateEvent.ComponentInstance.Id, 10)
+		serviceId := strconv.FormatInt(updateEvent.ComponentInstance.ServiceId, 10)
+		componentVersionId := strconv.FormatInt(updateEvent.ComponentInstance.ComponentVersionId, 10)
+
+		// Update service relation
+		removeServiceInput := openfga.RelationInput{
```

Comment:
Not every update will change the `serviceId` or `componentVersionId`. Can we check if has been changed and only do this operations if required?